File: BaekJoon/DFS, BFS/7562_나이트의 이동.py
def bfs(visited, srt_x, srt_y, dst_x, dst_y, size):
    queue = []

    dx = [-2, -1, 1, 2, 2, 1, -1, -2] # 가능한 방향
    dy = [1, 2, 2, 1, -1, -2, -2, -1]

    queue.append((srt_x, srt_y))

    while queue:
        element = queue.pop(0)
        x = element[0]
        y = element[1]

        if x == dst_x and y == dst_y: # 목적지에 도착한 경우
            return visited[x][y]

        for i in range(8):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx >= 0 and nx < size and ny >= 0 and ny < size: # 이동하려고 하는 좌표가 범위 이내일 때
                if visited[nx][ny] == 0:    # 방문하지 않은 좌표일 때
                    visited[nx][ny] = visited[x][y] + 1 # 이전 좌표까지의 횟수 + 1
                    queue.append((nx, ny))

def main():
    n = int(input())

    for _ in range(n):
        size = int(input())
        visited = [[0 for _ in range(size)] for _ in range(size)]
        srt_x, srt_y = map(int, input().split())
        dst_x, dst_y = map(int, input().split())
        print(bfs(visited, srt_x, srt_y, dst_x, dst_y, size))

if __name__ == "__main__":
    main()
    

File: BaekJoon/DFS, BFS/7562_나이트의 이동.cpp
#include <iostream>
#include <queue>
using namespace std;

int visit[300][300];
int dx[8] = {-2, -1, 1, 2, 2, 1, -1, -2};
int dy[8] = {1, 2, 2, 1, -1, -2, -2, -1};
int n, m, srt_x, srt_y, dst_x, dst_y;

int bfs(){
    int x, y, nx, ny;
    queue <pair <int, int>> Q;
    Q.push({srt_x, srt_y});

    while (Q.size() > 0){
        x = Q.front().first;
        y = Q.front().second;
        Q.pop(); // 첫 번째 원소 제거
        
        if (x == dst_x && y == dst_y){ // 도착지인 경우
            return visit[x][y];
        } 

        for (int i = 0; i < 8; i++){ // 모든 방향 탐색
            nx = x + dx[i];
            ny = y + dy[i];

            if ((nx >= 0 && nx < m) && (ny >= 0 && ny < m)){ // 이동하려고 하는 좌표가 범위 이내일 때
                if (visit[nx][ny] == 0){ // 아직 방문하지 않은 좌표일 때
                    visit[nx][ny] = visit[x][y] + 1; // 이전 좌표까지의 횟수 + 1
                    Q.push({nx, ny});
                }
            }
        }
    }
}

int main(){
    cin >> n;

    for(int i = 0; i < n; i++){
        cin >> m >> srt_x >> srt_y >> dst_x >> dst_y;
        for (int j = 0; j < m; j++){
            for (int k = 0; k < m; k++){
                visit[j][k] = 0;
            }
        }
        cout << bfs() << endl;
    }
}